Drop replicas from dirname for internal_replication=true

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Drop replicas from dirname for internal_replication=true (allows INSERT into Distributed with cluster from any number of replicas, before only 15 replicas was supported, everything more will fail with ENAMETOOLONG while creating directory for async blocks)

Detailed description / Documentation draft:
Under use_compact_format_in_distributed_parts_names=1 and
internal_replication=true the server encodes all replicas for the
directory name for async INSERT into Distributed, and the directory name
looks like:

    shard1_replica1,shard1_replica2,shard1_replica3

This is required for creating connections (to specific replicas only),
but in case of internal_replication=true, this can be avoided, since
this path will always includes all replicas.

This patch replaces all replicas with "_all_replicas" marker.

Note, that initial problem was that this path may overflow the NAME_MAX
if you will have more then 15 replicas, and the server will fail to
create the directory.

Also note, that changed directory name should not be a problem, since:
- empty directories will be removed since #16729
- and replicas encoded in the directory name is also supported anyway.

Follow-up for: #9653

alexey-milovidov commented:
Sorry, I don't understand, where 16 comes from and how `shardN_replicaM` does not support infinite number of nodes, e.g. `shard9999_replica9999`

azat commented:
>Sorry, I don't understand, where 16 comes from and how shardN_replicaM does not support infinite number of nodes, e.g. shard9999_replica9999

Since for async INSERT (`insert_distributed_sync=0`) and `internal_replication=true` it creates directory that includes all replicas in the name, i.e. `/var/lib/clickhouse/data/default/dist_data/shard10_replica1,shard10_replica2,shard10_replica3,shard10_replica4,shard10_replica5,shard10_replica6,shard10_replica7,shard10_replica8,shard10_replica9/`

And it is limited by 255 (`NAME_MAX`) on linux, so:

```
# use_compact_format_in_distributed_parts_names=2, this patch
In [3]: len(','.join([f's{i}_r{i}' for i in range(34)]))
Out[3]: 251

# use_compact_format_in_distributed_parts_names=1
In [4]: len(','.join([f'shard{i}_replica{i}' for i in range(15)]))
Out[4]: 249
```

alexey-milovidov commented:
> it creates directory that includes all replicas in the name

It is a mistake, should be simply `shard10`.
Let's correct this mistake.

azat committed:
Drop replicas from dirname for internal_replication=true

Under use_compact_format_in_distributed_parts_names=1 and
internal_replication=true the server encodes all replicas for the
directory name for async INSERT into Distributed, and the directory name
looks like:

    shard1_replica1,shard1_replica2,shard3_replica3

This is required for creating connections (to specific replicas only),
but in case of internal_replication=true, this can be avoided, since
this path will always includes all replicas.

This patch replaces all replicas with "_all_replicas" marker.

Note, that initial problem was that this path may overflow the NAME_MAX
if you will have more then 15 replicas, and the server will fail to
create the directory.

Also note, that changed directory name should not be a problem, since:
- empty directories will be removed since #16729
- and replicas encoded in the directory name is also supported anyway.

azat commented:
>Let's correct this mistake.

Then we cannot know what replicas had been there for creating connections - https://github.com/ClickHouse/ClickHouse/blob/bf482f75a708a10fea8657f26ef5bc8756135b72/src/Storages/Distributed/DirectoryMonitor.cpp#L67-L69

But let's add a special marker, i.e. `shardX_all_replicas` that maybe used to create connections to all replicas, this is fine for `internal_replication=true`.

azat commented:
[Functional stateless tests (memory) — fail: 1, passed: 3132, skipped: 14](https://clickhouse-test-reports.s3.yandex.net/25513/b96df135aee40b1a54a4fc3f0d4db89e11385564/functional_stateless_tests_(memory).html#fail1)

#25605

azat commented:
@alexey-milovidov should it be rebased or existing passed checks is enough?
